<launch>
    <!-- Arguments -->
    <arg name="mode_master" default="false" />  <!-- Master arm initial mode -->
    <arg name="auto_enable_slave" default="true" />

    <!-- Master arms: mode=0, publish control commands -->
    <!-- Left master -->
    <node name="$(anon piper_master_left)" 
          pkg="piper" 
          type="piper_start_ms_node_new.py" 
          output="screen">
        <param name="can_port" value="can_left_mas"/>
        <param name="mode" value="0"/>  <!-- Read and publish control -->
        <param name="auto_enable" value="false"/>
        <param name="mode_master" value="$(arg mode_master)"/>

        <remap from="/puppet/arm_status" to="/puppet_master/arm_status_left" />
        <remap from="/puppet/joint_states" to="/puppet_master/joint_left" />
        <remap from="/master/joint_states" to="/master/joint_left" />
        <remap from="/puppet/end_pose" to="/puppet_master/end_pose_left" />
        <remap from="/pos_cmd" to="/puppet_master/pos_cmd_left" />
        <remap from="/puppet/end_pose_euler" to="/puppet_master/end_pose_euler_left" />

        <remap from="/master/arm_status" to="/teach/master_control_status_left" />
        <remap from="/master/mode_status" to="/teach/master_mode_status_left" />
        <remap from="/master/enable" to="/teach/master_enable_left" />
        <remap from="/master/linkage_config" to="/teach/master_config_left" />
        <remap from="/master/teach_mode" to="/teach/teach_mode_left" />
        <remap from="/master_controled/joint_states" to="/master_controled/joint_left" />
    </node>

    <!-- Right master -->
    <node name="$(anon piper_master_right)" 
          pkg="piper" 
          type="piper_start_ms_node_new.py" 
          output="screen">
        <param name="can_port" value="can_right_mas"/>
        <param name="mode" value="0"/>
        <param name="auto_enable" value="false"/>
        <param name="mode_master" value="$(arg mode_master)"/>

        <remap from="/puppet/arm_status" to="/puppet_master/arm_status_right" />
        <remap from="/puppet/joint_states" to="/puppet_master/joint_right" />
        <remap from="/master/joint_states" to="/master/joint_right" />
        <remap from="/puppet/end_pose" to="/puppet_master/end_pose_right" />
        <remap from="/pos_cmd" to="/puppet_master/pos_cmd_right" />
        <remap from="/puppet/end_pose_euler" to="/puppet_master/end_pose_euler_right" />

        <remap from="/master/arm_status" to="/teach/master_control_status_right" />
        <remap from="/master/mode_status" to="/teach/master_mode_status_right" />
        <remap from="/master/enable" to="/teach/master_enable_right" />
        <remap from="/master/linkage_config" to="/teach/master_config_right" />
        <remap from="/master/teach_mode" to="/teach/teach_mode_right" />
        <remap from="/master_controled/joint_states" to="/master_controled/joint_right" />
    </node>

    <!-- Slave arms: mode=1, subscribe and execute control commands -->
    <!-- Left slave -->
    <node name="$(anon piper_slave_left)" 
          pkg="piper" 
          type="piper_start_ms_node_new.py" 
          output="screen">
        <param name="can_port" value="can_left_slave"/>
        <param name="mode" value="1"/>  <!-- Slave: subscribe and execute -->
        <param name="auto_enable" value="$(arg auto_enable_slave)"/>

        <remap from="/puppet/arm_status" to="/puppet/arm_status_left" />
        <remap from="/puppet/joint_states" to="/puppet/joint_left" />
        <remap from="/master/joint_states" to="/master/joint_left" />
        <remap from="/puppet/end_pose" to="/puppet/end_pose_left" />
        <remap from="/pos_cmd" to="/puppet/pos_cmd_left" />
        <remap from="/puppet/end_pose_euler" to="/puppet/end_pose_euler_left" />
        <remap from="/enable_flag" to="/puppet/enable_left" />
    </node>

    <!-- Right slave -->
    <node name="$(anon piper_slave_right)" 
          pkg="piper" 
          type="piper_start_ms_node_new.py" 
          output="screen">
        <param name="can_port" value="can_right_slave"/>
        <param name="mode" value="1"/>
        <param name="auto_enable" value="$(arg auto_enable_slave)"/>

        <remap from="/puppet/arm_status" to="/puppet/arm_status_right" />
        <remap from="/puppet/joint_states" to="/puppet/joint_right" />
        <remap from="/master/joint_states" to="/master/joint_right" />
        <remap from="/puppet/end_pose" to="/puppet/end_pose_right" />
        <remap from="/pos_cmd" to="/puppet/pos_cmd_right" />
        <remap from="/puppet/end_pose_euler" to="/puppet/end_pose_euler_right" />
        <remap from="/enable_flag" to="/puppet/enable_right" />
    </node>

</launch>
